lua/telescope/sorters.lua

```diff
@@ -70,13 +71,229 @@ sorters.get_levenshtein_sorter = function()
+        local curr = point_list[idx]
+
+        if curr.area_idx == prev_area and
+          curr.area_offset == prev_offset + 1 then
+          if from_upper_case then
+            bonus_caps_matcher = bonus_caps_matcher + 1
+          end
+          cc = cc + 1
+        else
+          cc = 1
+          from_upper_case = curr.is_upper_case
+          bonus_caps_matcher = 0
+        end
+
+        score = score +
+        (cc + bonus_caps_matcher) * -- Using caps should be worth more in score
+        (1 +
+          curr.area_idx / area_count + -- More specific is worth more
+          cc / #areas[curr.area_idx].word + -- More percent of an area matched
+          (curr.exact_match and 0.1 or 0) + -- case matching
+          (curr.area_idx == area_count and 0.5 or 0)) -- A bonus for final word
+
+        prev_area = curr.area_idx
+        prev_offset = curr.area_offset
+      end
+
+      if out < score then
+        out = score
+      end
+    end
+
+    return out
+  end
+
+  local function find_points(areas, start_points, prompt, prompt_idx, exact_match)
+
+    local local_points = {}
+    local start_point = start_points[1]
+    local area_idx = start_point.area_idx
+    local area_offset = start_point.area_offset
+
+    repeat
+      local curr_area = areas[area_idx]
+      local target = prompt:byte(prompt_idx)
+
+      if target == os_sep then
+        -- I HATE THIS WTHTY ONTHOEUN YROEU NTOHEUNCFINTHINTHSN ONTTDH E MT
+        return start_points
+      end
+
+      for idx = area_offset + 1, #curr_area.word do
+        if target == curr_area.word:byte(idx) or
+          not exact_match and target - 32 == curr_area.word:byte(idx) then
+          table.insert(local_points, {
+              area_idx = area_idx,
+              area_offset = idx,
+              exact_match = target == curr_area.word:byte(idx),
+              is_upper_case = target >= 65 and target <= 90,
+            })
+        end
+      end
+
+      area_idx = area_idx + 1
+      area_offset = 0
+
+    until area_idx > #areas
+
+    local return_points = {}
+    for idx = 1, #start_points do
+      local point = start_points[idx]
+      for local_idx = 1, #local_points do
+        local local_point = local_points[local_idx]
+        if local_point.area_idx == point.area_idx and
+          local_point.area_offset > point.area_offset or
+          local_point.area_idx > point.area_idx then
+          table.insert(return_points, {
+              area_idx = local_point.area_idx,
+              area_offset = local_point.area_offset,
+              previous = point,
+              exact_match = local_point.exact_match,
+              is_upper_case = local_point.is_upper_case,
+            })
+        end
+      end
+    end
+
+    return return_points
+  end
+
+  local line_cache = {}
+  local function get_cache_for(line, prompt)
+    if not line_cache[line] then
+      line_cache[line] = Cache.PromptCache:new()
+    end
+
+    return line_cache[line]:get_cache(prompt)
+  end
```

Comment:
this strikes me as though we're going to blow up on memory for some of the larger searches. I'll have to ask you about this in our call.